plugins/aggregators/histogram/histogram.go

```diff
@@ -0,0 +1,239 @@
+type bucketsByFields map[string]buckets
+
+// buckets contains the right borders buckets
+type buckets []float64
+
+// metricHistogramCollection aggregates the histogram data
+type metricHistogramCollection struct {
+	histogramCollection map[string]counts
+	metric              string
+	tags                map[string]string
+}
+
+// counts is the number of hits in the bucket
+type counts []uint64
+
+// NewHistogramAggregator creates new histogram aggregator
+func NewHistogramAggregator() telegraf.Aggregator {
+	h := &HistogramAggregator{}
+	h.buckets = make(bucketsByMetrics)
+	h.resetCache()
+
+	return h
+}
+
+// sampleConfig is config sample of histogram aggregation plugin
+var sampleConfig = `
+  # # Configuration for aggregate histogram metrics
+  # [[aggregators.histogram]]
+  #   ## General Aggregator Arguments:
+  #   ## The period on which to flush & clear the aggregator.
+  #   period = "30s"
+  #   ## If true, the original metric will be dropped by the
+  #   ## aggregator and will not get sent to the output plugins.
+  #   drop_original = false
+  #
+  #   ## The example of config to aggregate histogram for all fields of specified metric.
+  #   [[aggregators.histogram.config]]
+  #     ## The set of buckets.
+  #     buckets = [0.0, 15.6, 34.5, 49.1, 71.5, 80.5, 94.5, 100.0]
+  #     ## The name of metric.
+  #     metric_name = "cpu"
+  #
+  #   ## The example of config to aggregate for specified fields of metric.
+  #   [[aggregators.histogram.config]]
+  #     ## The set of buckets.
+  #     buckets = [0.0, 10.0, 20.0, 30.0, 40.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0]
+  #     ## The name of metric.
+  #     metric_name = "system"
+  #     ## The concrete field of metric
+  #     metric_field = "load1"
+`
+
+// SampleConfig returns sample of config
+func (h *HistogramAggregator) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns description of aggregator plugin
+func (h *HistogramAggregator) Description() string {
+	return "Keep the aggregate histogram of each metric passing through."
+}
+
+// Add adds new hit to the buckets
+func (h *HistogramAggregator) Add(in telegraf.Metric) {
+	id := in.HashID()
+	agr, ok := h.cache[id]
+	if !ok {
+		agr = metricHistogramCollection{
+			metric:              in.Name(),
+			tags:                in.Tags(),
+			histogramCollection: make(map[string]counts),
+		}
+	}
+
+	for field, value := range in.Fields() {
+		buckets := h.getBuckets(in.Name(), field)
+		if buckets == nil {
+			continue
+		}
+
+		if agr.histogramCollection[field] == nil {
+			agr.histogramCollection[field] = make(counts, len(buckets)+1)
+		}
+
+		if value, ok := convert(value); ok {
+			index := sort.SearchFloat64s(buckets, value)
+			if index < len(agr.histogramCollection[field]) {
+				agr.histogramCollection[field][index]++
+			}
+		}
+	}
+
+	h.cache[id] = agr
+}
+
+// Push returns histogram values for metrics
+func (h *HistogramAggregator) Push(acc telegraf.Accumulator) {
+	var isResetNeeded = false
```

Comment:
yes, that works, although you should change the counts to be int64 instead of uint64, the reason being that influxdb won't accept numbers over int64 (I believe prometheus may accept numbers over this, but you will begin to get odd rounding behavior)